include/LinearAlgebra/sparseLUSolver.h

```diff
@@ -192,6 +192,21 @@ void SparseLUSolver<T>::factorizeWithHashing(const SparseMatrixCSR<T>& A)
             else
                 U_map[i][j] = val;
         }
+
+        const T tolerance_abs = 1e-12; // minimum allowed diagonal
+        const T tolerance_rel = 1e-8; // relative to the max in the row
+
+        // Static pivoting / diagonal perturbation
+        T& diag   = U_map[i][i]; // reference to the diagonal
+        T max_val = 0;
+        for (const auto& [col, val] : U_map[i]) {
+            max_val = std::max(max_val, std::abs(val));
+        }
+
+        T threshold = std::max(tolerance_abs, tolerance_rel * max_val);
+        if (std::abs(diag) < threshold) {
+            diag = (diag >= 0) ? threshold : -threshold;
```

Comment:
(copysign)[https://en.cppreference.com/w/cpp/numeric/math/copysign] may be faster than an if condition
```suggestion
            diag = copysign(threshold, diag);
```